README
NAME
    RT::Extension::ACNS - parse ACNS messages and extract info into custom
    fields

DESCRIPTION
    ACNS stands for Automated Copyright Notice System. It's an open source,
    royalty free system that universities, ISP's, or anyone that handles
    large volumes of copyright notices can implement on their network to
    increase the efficiency and reduce the costs of responding to the
    notices... See <http://mpto.unistudios.com/xml/> for more details.

    This extension for RT is a configurable scrip that parses ACNS XML from
    incomming messages and stores it in custom fields.

RT VERSION
    Works with RT 4.0, RT 4.2 and 5.0.

INSTALLATION
    perl Makefile.PL
    make
    make install
        May need root permissions

    make initdb
        Only run this the first time you install this module.

        If you run this twice, you may end up with duplicate data in your
        database.

        If you are upgrading this module, check for upgrading instructions
        in case changes need to be made to your database.

    Edit your /opt/rt5/etc/RT_SiteConfig.pm
        If you are using RT 4.2 or greater, add this line:

            Plugin('RT::Extension::ACNS');

        For RT 4.0, add this line:

            Set(@Plugins, qw(RT::Extension::ACNS));

        or add RT::Extension::ACNS to your existing @Plugins line.

    Configure ACNS
        The scrip is configured through the %ACNS config option, as
        described in etc/RT_ACNSConfig.pm.

    Restart your webserver

AUTHOR
    Best Practical Solutions, LLC <modules@bestpractical.com>

BUGS
    All bugs should be reported via email to

        L<bug-RT-Extension-ACNS@rt.cpan.org|mailto:bug-RT-Extension-ACNS@rt.cpan.org>

    or via the web at

        L<rt.cpan.org|http://rt.cpan.org/Public/Dist/Display.html?Name=RT-Extension-ACNS>.

LICENSE AND COPYRIGHT
    This software is Copyright (c) 2014 by Best Practical Solutions

    This is free software, licensed under:

      The GNU General Public License, Version 2, June 1991



Changes
1.01 2023-07-24
 - Update Module::Install files
 - Update POD mentioning it is compatible with RT 5.0

1.00 2014-12-17
 - Packaging and documentation updates

0.04 2013-04-25
 - add support for functions in the map

0.03 2013-04-11
 - extract from a verified emails (gpg strips some bits)

0.02 2013-03-26
 - extract from XML in an attachment

0.01 2011-02-04
 - initial release



lib/RT/Extension/ACNS/Test.pm
use strict;
use warnings;

### after: use lib qw(@RT_LIB_PATH@);
use lib qw(/opt/rt5/local/lib /opt/rt5/lib);

package RT::Extension::ACNS::Test;

our @ISA;
BEGIN {
    local $@;
    eval { require RT::Test; 1 } or do {
        require Test::More;
        Test::More::BAIL_OUT(
            "requires 3.8 to run tests. Error:\n$@\n"
            ."You may need to set PERL5LIB=/path/to/rt/lib"
        );
    };
    push @ISA, 'RT::Test';
}

sub import {
    my $class = shift;
    my %args  = @_;

    $args{'requires'} ||= [];
    if ( $args{'testing'} ) {
        unshift @{ $args{'requires'} }, 'RT::Extension::ACNS';
    } else {
        $args{'testing'} = 'RT::Extension::ACNS';
    }

    $class->SUPER::import( %args );
    $class->export_to_level(1);

    $RT::Config::META{'ACNS'} = { Type => 'HASH' };

    require RT::Extension::ACNS;
}

1;


etc/RT_ACNSConfig.pm
=head1 RT::Extension::ACNS specific settings

=head2 %ACNS

A hash with Defaults and Map keys.

Defaults are used for every ACNS message, it's a hash
with ( custom field => value ) pairs. For example:

    Set( %ACNS => 
        Defaults => {
            'Custom Field X' => 'default value',
            ...
        },
        Map => ...
    );

Map is hash with (custom field => path) pairs, where path
is an array reference with node names from ACNS XML. More
on paths below.

    Set( %ACNS => 
        Defaults => { ... },
        Map => {
            'A custom field' => ['element', 'element', 'element'],
            ...
        },
    );

Path in the map follows XML structure strictly with some tricks.

=over 4

=item * don't need to list Infringement as first element, for example:

    # store text of Infringement/Case/ID in Case custom field
    Set( %ACNS => 
        Map => {
            Case => [qw(Case ID)],
        },
    );

=item * ACNS specification has repeated elements, for example there may be
multiple Item records in Infringement/Content section. Such elements should
be followed by '1', '-1' or '*' to pick first, last or all sub-items
correspondingly. For example:

    # store all URLs in URLs custom field
    Set( %ACNS => 
        Map => {
            URLs => [qw(Content Item * URL)],
        },
    );

If you picked all elements then multiple values stored in the custom field,
but if the CF can store only one value then values joined with new line
character before storing.

=item * a few elements in ACNS schema has attributes and text. Follow element
name with attribute name to store text. For example:

    # store last notice timestamp
    Set( %ACNS => 
        Map => {
            LastNotice => [qw(History Notice -1 TimeStamp)],
        },
    );

=back

=cut

Set( %ACNS => 
    Defaults => { },
    Map => { },
);


lib/RT/Action/ParseACNS.pm
use strict;
use warnings;

package RT::Action::ParseACNS;
use base 'RT::Action';

use XML::LibXML;
use Parse::ACNS;

sub Prepare {
    my $self = shift;

    my $xml;

    my $txn = $self->TransactionObj;

    my $content = $txn->Content;
    if ( $content =~ s/^.*Start ACNS XML\n//s ) {
        $content =~ s/(- )?-+End ACNS XML.*$//s;
        $xml = XML::LibXML->new->parse_string( $content );
    } else {
        my $attachments = $txn->Attachments;
        while ( my $attach = $attachments->Next ) {
            next unless ($attach->Filename||'') =~ /\.xml$/;
            next unless ($attach->Content||'') =~ /Infringement/i;

            $xml = XML::LibXML->new->parse_string( $attach->Content );
            last;
        }
    }
    unless ($xml) {
        $RT::Logger->error("Failed to find ACNS XML");
        return 1;
    }

    my $data = Parse::ACNS->new->parse( $xml );
    $self->{'ACNS'} = $self->MapDataOverCFs( $data );
    return 1;
}

sub Commit {
    my $self = shift;
    
    $self->UpdateCustomFields( $self->{'ACNS'} );
    return 1;
}

sub UpdateCustomFields {
    my $self = shift;
    my $values = shift;
    while ( my ($name, $value) = each %$values ) {
        my $cf = $self->TicketObj->LoadCustomFieldByIdentifier( $name );
        unless ( $cf && $cf->id ) {
            $RT::Logger->error( "Ticket #". $self->TicketObj->id . "has no custom field '$name'");
            next;
        }
        $self->UpdateCustomField( $cf, $value);
    }
}

sub UpdateCustomField {
    my $self = shift;
    my $cf = shift;
    my $value = shift;

    my $ticket = $self->TicketObj;

    if ( $cf->MaxValues == 1 ) {
        if ( $value ) {
            if ( ref $value ) {
                $RT::Logger->debug(
                    "Custom Field can have only one value, but we got"
                    ." several values from ACNSS data. Joining with newlines."
                );
                $value = join "\n", @$value;
            }
            my ($status, $msg) = $ticket->AddCustomFieldValue(
                Field => $cf, Value => $value,
            );
            $RT::Logger->error("Couldn't set CF: $msg") unless $status;
        }
        elsif ( $ticket->CustomFieldValues->First ) {
            my ($status, $msg) = $ticket->DeleteCustomFieldValue(
                Field => $cf,
                Value => $ticket->FirstCustomFieldValue( $cf ),
            );
            $RT::Logger->error("Couldn't delete CF value: $msg")
                unless $status;
        }
    }
    else {
        unless ( $value ) {
            foreach my $value ( @{ $ticket->CustomFieldValues($cf)->ItemsArrayRef } ) {
                my ($status, $msg) = $ticket->DeleteCustomFieldValue(
                    Field   => $cf, ValueId => $value->id,
                );
                $RT::Logger->error("Couldn't delete CF value: $msg")
                    unless $status;
            }
        }
        else {
            my @new = ref $value? (@$value) : ($value);
            my @old = @{ $ticket->CustomFieldValues($cf)->ItemsArrayRef };

            my @tmp;
            foreach my $new ( @new ) {
                next if grep lc $_->Content eq $new, @old;
                push @tmp, $new;
            }
            foreach my $old ( splice @old ) {
                my $oldv = lc $old->Content;
                next unless grep $oldv eq lc $_, @new;
                push @old, $old;
            }
            @new = @tmp;

            foreach my $value ( @old ) {
                my ($status, $msg) = $ticket->DeleteCustomFieldValue(
                    Field   => $cf, ValueId => $value->id,
                );
                $RT::Logger->error("Couldn't delete CF value: $msg")
                    unless $status;
            }
            foreach my $value ( @new ) {
                my ($status, $msg) = $ticket->AddCustomFieldValue(
                    Field   => $cf, Value => $value,
                );
                $RT::Logger->error("Couldn't add CF value: $msg")
                    unless $status;
            }
        }
    }
}

sub MapDataOverCFs {
    my $self = shift;
    my $data = shift;

    my %config = RT->Config->Get('ACNS');

    my %res;
    %res = %{ $config{'Defaults'} } if $config{'Defaults'};
    return \%res unless $config{'Map'};

    while ( my ($cf, $path) = each %{ $config{'Map'} } ) {
        my @tmp = grep defined && length, $self->ResolveMapEntry(
            Data => $data,
            Path => $path,
            CustomField => $cf,
        );
        $res{ $cf } = @tmp > 1 ? [ @tmp ] : @tmp? $tmp[0] : undef;
    }
    return \%res;
}

sub ResolveMapEntry {
    my $self = shift;
    my %args = @_;

    my $data = $args{'Data'};
    my @path = @{ $args{'Path'} };
    my @done = @{ $args{'Done'} || [] };

    $RT::Logger->debug("Searching for '". join('.', @path) ."' in ACNS data" )
        unless @done; # log once

    while ( my $e = shift @path ) {
        unless ( ref $data ) {
            $RT::Logger->error("Reached terminal element in data");
            return ();
        }
        elsif ( 'HASH' eq ref $data ) {
            unless ( exists $data->{ $e } ) {
                $RT::Logger->debug(
                    "No entry for '". join('.', @done, $e)
                    ."' in ACNS data. Ignoring"
                );
                return ();
            }
            else {
                push @done, $e;
                $data = $data->{ $e };
            }
        }
        elsif ( 'ARRAY' eq ref $data ) {
            if ( $e eq '*' ) {
                push @done, $e;
                my @res;
                foreach my $data_point ( @{$data} ) {
                    push @res, $self->ResolveMapEntry(
                        Data => $data_point,
                        Path => [ @path ],
                        Done => \@done,
                    );
                }
                return @res;
            }
            elsif ( $e eq '1' || $e eq '-1' ) {
                push @done, $e;
                $data = $data->[ $e > 0? 0 : -1 ];
            }
            else {
                $RT::Logger->error(
                    "Reached list '". join('.', @done)
                    ."', but selector ($e) is not *, 1 or -1"
                );
                return ();
            }
        }
    }

    unless ( ref $data ) {
        return ($data);
    }
    elsif ( 'HASH' eq ref $data && exists $data->{'_'} ) {
        return ($data->{'_'});
    }
    else {
        $RT::Logger->error(
            "Found '". join('.', @done)
            ."', but it's not an end of the ACNS data"
        );
        return ();
    }
}

1;


lib/RT/Extension/ACNS.pm
use 5.008003;
use strict;
use warnings;

package RT::Extension::ACNS;

our $VERSION = '1.01';

=head1 NAME

RT::Extension::ACNS - parse ACNS messages and extract info into custom fields

=head1 DESCRIPTION

ACNS stands for Automated Copyright Notice System. It's an open source,
royalty free system that universities, ISP's, or anyone that handles
large volumes of copyright notices can implement on their network to
increase the efficiency and reduce the costs of responding to the
notices...  See L<http://mpto.unistudios.com/xml/> for more details.

This extension for RT is a configurable scrip that parses ACNS XML from
incomming messages and stores it in custom fields.

=head1 RT VERSION

Works with RT 4.0, RT 4.2 and 5.0.

=head1 INSTALLATION

=over

=item C<perl Makefile.PL>

=item C<make>

=item C<make install>

May need root permissions

=item C<make initdb>

Only run this the first time you install this module.

If you run this twice, you may end up with duplicate data
in your database.

If you are upgrading this module, check for upgrading instructions
in case changes need to be made to your database.

=item Edit your F</opt/rt5/etc/RT_SiteConfig.pm>

If you are using RT 4.2 or greater, add this line:

    Plugin('RT::Extension::ACNS');

For RT 4.0, add this line:

    Set(@Plugins, qw(RT::Extension::ACNS));

or add C<RT::Extension::ACNS> to your existing C<@Plugins> line.

=item Configure ACNS

The scrip is configured through the C<%ACNS> config option, as described
in F<etc/RT_ACNSConfig.pm>.

=item Restart your webserver

=back

=head1 AUTHOR

Best Practical Solutions, LLC E<lt>modules@bestpractical.comE<gt>

=head1 BUGS

All bugs should be reported via email to

    L<bug-RT-Extension-ACNS@rt.cpan.org|mailto:bug-RT-Extension-ACNS@rt.cpan.org>

or via the web at

    L<rt.cpan.org|http://rt.cpan.org/Public/Dist/Display.html?Name=RT-Extension-ACNS>.

=head1 LICENSE AND COPYRIGHT

This software is Copyright (c) 2014 by Best Practical Solutions

This is free software, licensed under:

  The GNU General Public License, Version 2, June 1991

=cut

1;
